sdk/src/main/java/com/horizen/cryptolibprovider/CswCircuitImplZendoo.java

```diff
@@ -0,0 +1,217 @@
+package com.horizen.cryptolibprovider;
+
+import com.horizen.block.WithdrawalEpochCertificate;
+import com.horizen.box.Box;
+import com.horizen.certnative.BackwardTransfer;
+import com.horizen.certnative.WithdrawalCertificate;
+import com.horizen.cswnative.CswFtProverData;
+import com.horizen.cswnative.CswProof;
+import com.horizen.cswnative.CswSysData;
+import com.horizen.cswnative.CswUtxoProverData;
+import com.horizen.fwtnative.ForwardTransferOutput;
+import com.horizen.librustsidechains.Constants;
+import com.horizen.librustsidechains.FieldElement;
+import com.horizen.proposition.Proposition;
+import com.horizen.provingsystemnative.ProvingSystemType;
+import com.horizen.scutxonative.ScUtxoOutput;
+import com.horizen.secret.PrivateKey25519;
+import com.horizen.utils.BytesUtils;
+import com.horizen.utils.ForwardTransferCswData;
+import com.horizen.merkletreenative.MerklePath;
+import com.horizen.utils.UtxoCswData;
+import com.horizen.utils.WithdrawalEpochUtils;
+
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class CswCircuitImplZendoo implements CswCircuit {
+
+    @Override
+    public int utxoMerkleTreeHeight() {
+        return Constants.SC_MST_HEIGHT();
+    }
+
+    @Override
+    public FieldElement getUtxoMerkleTreeLeaf(Box<Proposition> box) {
+        ScUtxoOutput utxo = new ScUtxoOutput(box.proposition().bytes(), box.value(), box.nonce(), box.customFieldsHash());
+        return utxo.getNullifier();
+    }
+
+    @Override
+    public byte[] getCertDataHash(WithdrawalEpochCertificate cert) {
+        WithdrawalCertificate we = createWithdrawalCertificate(cert);
+
+        FieldElement hashFe = we.getHash();
+        byte[] hashBytes = hashFe.serializeFieldElement();
```

Comment:
missing a try finally for hashFe.freeFieldElement(); ?

Reply:
Seems that `freeFieldElement` doesn't throw exceptions. It does nothing in case was freed before

Reply:
the additional try finally was to ensure that hashFe.freeFieldElement is always executed even in case of any kind of error (including RuntimeExceptions) of hashFe.serializeFieldElement()